src/Simplex/Chat.hs

```diff
@@ -1675,13 +1736,34 @@ processAgentMessage (Just user@User {userId, profile}) agentConnId agentMessage
                   toView $ CRReceivedContactRequest cReq
                   showToast (localDisplayName <> "> ") "wants to connect to you"
 
-    withAckMessage :: ConnId -> MsgMeta -> m () -> m ()
-    withAckMessage cId MsgMeta {recipient = (msgId, _)} action =
-      action `E.finally` withAgent (\a -> ackMessage a cId msgId `catchError` \_ -> pure ())
-
-    ackMsgDeliveryEvent :: Connection -> MsgMeta -> m ()
-    ackMsgDeliveryEvent Connection {connId} MsgMeta {recipient = (msgId, _)} =
-      withStore $ \db -> createRcvMsgDeliveryEvent db connId msgId MDSRcvAcknowledged
+    withCompletedCommand :: Connection -> ACommand 'Agent -> (CommandData -> m ()) -> m ()
+    withCompletedCommand Connection {connId} agentMsg action = do
+      cmdData_ <- withStore' $ \db -> getCommandDataByCorrId db user corrId
+      forM_ cmdData_ $ \cmdData@CommandData {cmdId, cmdConnId, cmdFunction} -> do
+        case cmdConnId of
+          Just cmdConnId' ->
+            if connId == cmdConnId'
+              then
+                if aCommandTag agentMsg == commandExpectedResponse cmdFunction
+                  then do
+                    withStore' $ \db -> updateCommandStatus db user cmdId CSCompleted
+                    action cmdData
+                  else throwChatError $ CECommandError "unexpected response for correlated command"
+              else throwChatError $ CECommandError "correlated command's connection id doesn't match"
+          Nothing -> throwChatError $ CECommandError "correlated command doesn't have connection id"
+
+    createAckCmd :: Connection -> m CommandId
+    createAckCmd Connection {connId} = do
+      withStore' $ \db -> createCommand db user (Just connId) CFAckMessage
+
+    withAckMessage :: ConnId -> CommandId -> MsgMeta -> m () -> m ()
+    withAckMessage cId cmdId MsgMeta {recipient = (msgId, _)} action =
+      -- [async agent commands] command should be asynchronous, continuation is ackMsgDeliveryEvent
+      action `E.finally` withAgent (\a -> ackMessageAsync a (aCorrId cmdId) cId msgId `catchError` \_ -> pure ())
```

Comment:
ackMessageAsync is non-trowing, right?

Reply:
I think it is, it reads connection and can throw CONN and CMD errors